compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,939 @@
+// For more information about type metadata and type metadata identifiers for cross-language LLVM
+// CFI support, see Type metadata in the design document in the tracking issue #89653.
+
+// FIXME(rcvalle): Identify C char and integer type uses and encode them with their respective
+// builtin type encodings as specified by the Itanium C++ ABI for extern function types with the "C"
+// calling convention to use this encoding for cross-language LLVM CFI.
+
+use bitflags::bitflags;
+use core::fmt::Display;
+use rustc_data_structures::base_n;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir as hir;
+use rustc_middle::ty::subst::{GenericArg, GenericArgKind, SubstsRef};
+use rustc_middle::ty::{
+    self, Binder, Const, ExistentialPredicate, FloatTy, FnSig, IntTy, List, Region, RegionKind,
+    Term, Ty, TyCtxt, UintTy,
+};
+use rustc_span::def_id::DefId;
+use rustc_target::abi::call::{Conv, FnAbi};
+use rustc_target::spec::abi::Abi;
+
+/// Type and extended type qualifiers.
+#[derive(Eq, Hash, PartialEq)]
+enum TyQ {
+    None,
+    Const,
+    Mut,
+}
+
+/// Substitution dictionary key.
+#[derive(Eq, Hash, PartialEq)]
+enum DictKey<'tcx> {
+    Ty(Ty<'tcx>, TyQ),
+    Region(Region<'tcx>),
+    Const(Const<'tcx>),
+    Predicate(ExistentialPredicate<'tcx>),
+}
+
+bitflags! {
+    /// Options for typeid_for_fnabi and typeid_for_fnsig.
+    pub struct TypeIdOptions: u32 {
+        const NO_OPTIONS = 0;
+        const GENERALIZE_POINTERS = 1;
+        const RESERVED = 2;
+    }
+}
+
+bitflags! {
+    /// Options for encode_ty.
+    struct EncodeTyOptions: u32 {
+        const NO_OPTIONS = 0;
+        const RESERVED = 1;
+        const GENERALIZE_REPR_C = 2;
+    }
+}
+
+bitflags! {
+    /// Options for transform_ty.
+    struct TransformTyOptions: u32 {
+        const NO_OPTIONS = 0;
+        const GENERALIZE_POINTERS = 1;
+        const GENERALIZE_REPR_C = 2;
+    }
+}
+
+/// Converts a number to a disambiguator (see
+/// https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html).
+fn to_disambiguator(num: u64) -> String {
+    if let Some(num) = num.checked_sub(1) {
+        format!("s{}_", base_n::encode(num as u128, 62))
+    } else {
+        "s_".to_string()
+    }
+}
```

Comment:
```suggestion
fn to_disambiguator(num: u64) -> Cow<'static, str>{
    if let Some(num) = num.checked_sub(1) {
        // FIXME: default size?
        let mut s = String::new();
        s.push('s');
        base_n::push_str(num as u128, 62, s);
        s.push('_');
        Cow::Owned(s)
    } else {
        Cow::Borrowed("s_")
    }
}
```

Reply:
Or maybe this should write to given buffer too?

Reply:
While I agree with the push_str+format optimization, we must be careful with micro-optimizations sacrificing API consistency and ergonomics, code readability and maintainability, but not bringing any significant benefits. I'll keep an eye out for and work on them if any issues arise.